src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,287 @@
+            uint32_t offset;
+            payload.read(plogId);
+            if (payload.getDataRemaining() > 0){
+                payload.read(offset);
+            }
+            else{
+                offset = 0;
+            }
+            dto::MetadataElement log_entry{.name=std::move(plogId), .offset=offset};
+            _streamLog.push_back(std::move(log_entry));
+        } 
+        payload.clear();
+    }
+    return _client.info(_streamLog[_streamLog.size()-1].name)
+    .then([this] (auto&& response){
+        auto& [status, resp] = response;
+        if (!status.is2xxOK()) {
+            throw std::runtime_error("unable to obtain the information of a plog");
+        }
+        _streamLog[_streamLog.size()-1].offset = std::get<0>(resp);
+        std::vector<seastar::future<std::tuple<Status, Payload> > > readFutures;
+        for (auto& log_entry: _streamLog){
+            readFutures.push_back(_client.read(log_entry.name, 0, log_entry.offset));
+        }
+        return seastar::when_all_succeed(readFutures.begin(), readFutures.end());
+    })
+    .then([this] (std::vector<std::tuple<Status, Payload> >&& responses){
+        std::vector<Payload> newPayloads;
+        for (auto& response: responses){
+            auto& [status, resp] = response;
+            if (!status.is2xxOK()){
+                throw std::runtime_error("unable to read a plog");
+            }
+            newPayloads.push_back(std::move(resp));
+        }
+        return seastar::make_ready_future<std::vector<Payload> >(std::move(newPayloads));
+    });
+}
+
+
+seastar::future<> 
+LogStream::write(Payload payload){
+    return _write(std::move(payload), 1);
+}
+
+seastar::future<> 
+LogStream::_write(Payload payload, uint32_t writeToWAL){
+    if (!_logStreamCreate){
+        K2INFO("Log stream does not create");
+        throw std::runtime_error("Log stream does not create");
+    }
+
+    String sealed_plogId;
+    String new_plogId;
+    uint32_t sealed_offest = 0;
+    
+    // determine whether the current plog will excceed the size limit after the incoming writing operation
+    if (_plogInfo[writeToWAL].second + payload.getSize() >= PLOG_MAX_SIZE){
+        sealed_plogId = _plogInfo[writeToWAL].first;
+        sealed_offest = _plogInfo[writeToWAL].second;
+        // change to the next available plog id
+        if (writeToWAL == 0){
+            new_plogId = _metadataPlogPool[_metadataPlogPool.size()-1];
+            _metadataPlogPool.pop_back();
+        }
+        else{
+            new_plogId = _walPlogPool[_walPlogPool.size()-1];
+            _walPlogPool.pop_back();
+        }
+        _plogInfo[writeToWAL].first = new_plogId;
+        _plogInfo[writeToWAL].second = 0;
+    }
+
+    uint32_t current_offset = _plogInfo[writeToWAL].second;
+    _plogInfo[writeToWAL].second += payload.getSize();
+
+    // write the incoming payload to current plog
+    return _client.append(_plogInfo[writeToWAL].first, current_offset,  std::move(payload))

```

Comment:
It seems like it is possible to leak plogs here, if we tried to allocate a new one from the pool above but the append call fails.

Reply:
I am not sure whether I need to handle the failure case here. If we want to handle this case, the only thing we can do is that, when we detected an error, we should seal all the opened Plogs?

Reply:
You don't need to fix it now but it would be good to add a TODO section to the design document. Another TODO is handling the case when the pool is empty which I see is an exception right now.